docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,960 @@
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. it is the legitimate holder of its node ID), _and_ that its attachments for the tenants
+  whose data is being deleted are current.
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
+   read metadata that does not reference the about-to-be-deleted object).
+3. Confirm we hold a non-stale generation for the node and the attachment(s) doing deletion.
+4. Actually delete the objects
+
+Note that at stage 3 we are only confirming that deletions of objects no longer referenced
+by the metadata written in step 2 are safe. While Step 3 is in flight, new deletions may
+be enqueued, but these are not safe to execute until a subsequent iteration of the
+generation check.
+
+Because step 3 puts load on the control plane, deletion lists should be executed lazily,
+accumulating many deletions before executing them behind a batched generation check message
+to the control plane, which would contain a list of tenant generation numbers to validate.
+
+The result of the request to the control plane may indicate that only some of the attachment
+generations are fresh: if this is the case, then the pageserver must selectively drop the
+deletions from a stale generation, but still execute the deletions for attachments
+that had a fresh generation.
+
+The only correctness requirement for deletions is that they are _not_ executed prior
+to the barrier in step 3. It is safe to delay a deletion, or indeed to never execute
+it at all, if a node restarts while an in-memory queue of deletions is pending. Minimizing
+leaking objects is an optimization, accomplished by making this queue persistent: see
+[Persistent Deletion Queue](#persistent-deletion-queue) in the optimizations section.
+
+### Deletion Part 2: WAL trim changes (delay `remote_consistent_lsn` updates)
+
+Remote objects are not the only kind of deletion the pageserver does: it also indirectly deletes
+WAL data, by feeding back remote_consistent_lsn to safekeepers, as a hint to the safekeepers that
+they may drop data below this LSN.
+
+We may solve for safety of remote_consistent_lsn updates in the same way as for deletions: before indicating to safekeepers
+that they may trim up to a particular LSN, pageservers should communicate with the control plane
+to confirm that their node generation is current. This guarantees that any future generation will
```

Comment:
This statement is only valid for a point in time though, no?

I mean, immediately after the control plane returned "yes, your generation is current", the control plane might bump that generation number and spin up a new pageserver for the node id. And while the pageserver with the old number that got the confirmation deletes data, the new pageserver with the new number starts up and does something that might race with the older pageserver's operations.

Reply:
The advertised LSN will be the one from before sending the validation request -- that way there is a strict ordering:
1. Write index_part
2. Stash remote_consistent_lsn value
3. RPC to control plane to confirm generation
4. Make remote_consistent_lsn value from step 2 visible to safekeepers

That series of steps is what provides the guarantee that when a safekeeper sees the lsn, one of the following is true:
- Original node is still latest genration
- or, a later node has taken over, but it must have seen the metadata from step 2, so its LSN is guaranteed to be >= the one that we stashed at step 2.

I need to adjust this language to make that clearer, currently it is a couple of paragraphs later:

> The `remote_consistent_lsn` advertised to a safekeeper is the value from _before_ the
validation request to the control plane started, not the value at the time the
pageserver receives a response. This will require an extra field internally to
the pageserver, to store some "pending remote_consistent_lsn" that may be shared
with the safekeeper after validation completes.